aesara/tensor/subtensor_opt.py

```diff
@@ -1666,3 +1668,100 @@ def local_subtensor_SpecifyShape_lift(fgraph, node):
         new_shape_arg = as_tensor(shape_arg[indices], dtype=np.int64, ndim=1)
 
     return [specify_shape(new_obj_arg, new_shape_arg)]
+
+
+def equal_idxs(idx1, idx2):
+    if isinstance(idx1, slice) and isinstance(idx2, slice):
+        return (
+            equal_computations([idx1.start], [idx2.start])
+            and equal_computations([idx1.stop], [idx2.stop])
+            and equal_computations([idx1.step], [idx2.step])
+        )
+    else:
+        return equal_computations([idx1], [idx2])
+
+
+@register_useless
+@register_canonicalize
+@register_specialize
+@local_optimizer([Join])
+def local_join_subtensors(fgraph, node):
+    """Simplify contiguous subtensors inside a join
+
+    join((x[:3], x[3:5]), axis=0) -> x[:5]
+    """
+    # TODO: Generalize to AdvancedSubtensors
+
+    axis, tensors = node.inputs[0], node.inputs[1:]
+
+    try:
+        axis = get_scalar_constant_value(axis)
+    except NotScalarConstantError:
+        return
+
+    for tensor1_idx, (tensor1, tensor2) in enumerate(zip(tensors[:-1], tensors[1:])):
+        # Check that two consecutive Subtensors are operating on the same base tensor
+        if not (
+            (tensor1.owner is not None and isinstance(tensor1.owner.op, Subtensor))
+            and (tensor2.owner is not None and isinstance(tensor2.owner.op, Subtensor))
+            and (tensor1.owner.inputs[0] is tensor2.owner.inputs[0])
+        ):
+            continue
+
+        # Check that subtensors have consecutive indexes across the join axis
+        idxs_tensor1 = indices_from_subtensor(
+            tensor1.owner.inputs[1:], tensor1.owner.op.idx_list
+        )
+        idxs_tensor2 = indices_from_subtensor(
+            tensor2.owner.inputs[1:], tensor2.owner.op.idx_list
+        )
+        try:
+            idxs_axis_tensor1 = idxs_tensor1[axis]
+            idxs_axis_tensor2 = idxs_tensor2[axis]
+        except IndexError:
+            continue
+        if isinstance(idxs_axis_tensor1, slice) and isinstance(
+            idxs_axis_tensor2, slice
+        ):
```

Comment:
If you `continue` on the negation of this condition, you can simplify these and the following statements.